### PR TITLE
NETOBSERV-1931: introduce flowcollector validation webhook

### PR DESCRIPTION
**Pull request by jotak**
## Description

Start by validating agent features

## Dependencies

n/a

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [x] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [ ] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [x] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

**Comment by jotak** [comment]
image

**Review by OlivierCazade — APPROVED** [review]
LGTM, 2 minor comments. This comments can be addressed later since this PR is blocking other work.

**Review by jpinsonneau — APPROVED** [review]
LGTM

From the user perspective, where does this warnings shows up in the console ?
It may be good to have these in the Status field :thinking:

File: apis/flowcollector/v1beta2/flowcollector_validation_webhook.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+			if err != nil {
+				warnings = append(warnings, fmt.Sprintf("Could not detect OpenShift cluster version: %s", err.Error()))
+			} else if !b {
+				warnings = append(warnings, fmt.Sprintf("The NetworkEvents feature requires OpenShift 4.18 or above (version detected: %s)", CurrentClusterInfo.GetOpenShiftVersion()))
+			}
+		} else {
+			warnings = append(warnings, "The NetworkEvents feature is only supported with OpenShift")
+		}
+		if !fc.Spec.Agent.EBPF.Privileged {
+			warnings = append(warnings, "The NetworkEvents feature requires eBPF Agent to run in privileged mode")
```
[the remark this answers is not in the document]
Comment (jotak) [reply apis/flowcollector/v1beta2/flowcollector_validation_webhook.go:98]: My rationale is that I'm setting warnings for things that won't work but won't break anything (e.g. a feature that will not be enabled), and errors for things that may break something or produce garbage / unexpected data (e.g. an invalid filter set on a metric will make the metric tell something completely different from the expectation)

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+			c.openShiftVersion = semver.New(history.Version)
+			break
+		}
+	}
+	c.fetchedClusterVersion = true
+	return nil
+}
+
+// MockOpenShiftVersion shouldn't be used except for testing
+func (c *Info) MockOpenShiftVersion(v string) {
```
[the remark this answers is not in the document]
Comment (jotak) [reply pkg/cluster/cluster.go:97]: it manipulates non-exported data (e.g. `apisMap`) so it has to be in the cluster package ; and it must be importable from other packages, so not in a `*_test.go` file ; we can't do that, right?

**Comment by jotak** [comment]
> LGTM
> 
> From the user perspective, where does this warnings shows up in the console ?

image

> It may be good to have these in the Status field 🤔

yeah I guess we can do that, let me create a follow-up task

**Comment by jotak** [comment]
follow-up: https://issues.redhat.com/browse/NETOBSERV-1938

**Pull request merged**